### PR TITLE
job: make eventlog contexts json objects instead of free form text

### PR DESCRIPTION
**Pull request by chu11**
Per discussion in #2076.  A few points worth mentioning:

- notes are no longer extra text, they are a key:value in the json object.  Typically listed with "note=..." format from `flux job`.
- `flux job eventlog` and `flux job wait-event` now have a `--context-format` option to output in raw json or nicer key=value output.  It defaults to key=value, so that most tests continue to work the same.
- when doing `json_dumps()` on a json string, it defaults to add quotes around the string.  In order to keep output identical to what was there before (with no quotes), I had to check if it was a string value first.

```
            json_object_foreach (o, key, value) {
                const char *sval;
                char *ptr = NULL;
                if (json_is_string (value))
                    sval = json_string_value (value);
                else {
                    ptr = json_dumps (value, JSON_ENCODE_ANY|JSON_COMPACT);
                    sval = ptr;
                }
                printf (" %s=%s", key, sval);
                free (ptr);
            }
```

ehhhh, maybe not worth it?  Could we just accept quotes around any strings?  Will have to adjust a number of tests as a result (i.e. `type=cancel` -> `type="cancel"`)
- I renamed `event_job_post_fmt()` to `event_job_post_pack()` b/c it now takes `json_pack()` like syntax for contexts.

### CODE REVIEW

**Comment by chu11** [comment]
looking through diff coverage at 75.96%, all important parts hit, just missed error paths

File: src/modules/job-manager/event.c [hunk 1]
```diff
@@ -240,29 +243,46 @@ int event_job_update (struct job *job, const char *event)
         if (job->state != FLUX_JOB_NEW)
             goto inval;
         job->t_submit = timestamp;
-        if (util_int_from_context (context, "priority", &priority) < 0)
+        if (!(o = json_loads (context, 0, NULL))) {
```
Comment (garlick) [inline src/modules/job-manager/event.c]: This function is going to get quite long as we add events.  Could we perhaps create functions to decode each event that has a context, e.g. for this one,  maybe
```c
int event_submit_context_decode (const char *event, int *priority, uint32_t *userid, int *flags);
```

File: src/modules/job-manager/event.c [hunk 2]
```diff
@@ -280,16 +300,20 @@ int event_job_update (struct job *job, const char *event)
...
+    save_errno = errno;
+    json_decref (o);
+    errno = save_errno;
     return -1;
 }
 
-int event_job_post (struct event_ctx *ctx, struct job *job,
-                    flux_continuation_f cb, void *arg,
-                    const char *name, const char *context)
+static int post (struct event_ctx *ctx, struct job *job,
```
Comment (garlick) [inline src/modules/job-manager/event.c]: What's the reason for this change?  It seems like we've created the `post()` function without gaining anything?

File: src/modules/job-manager/event.c [hunk 3]
```diff
@@ -316,22 +340,37 @@ int event_job_post (struct event_ctx *ctx, struct job *job,
...
-    if (n < 0 || n >= sizeof (context)) {
+    if (!(o = json_vpack_ex (NULL, 0, fmt, ap))) {
         errno = EINVAL;
-        return -1;
+        goto error;
     }
-    return event_job_post (ctx, job, cb, arg, name, context);
+    if (!(context = json_dumps (o, JSON_COMPACT))) {
+        errno = EINVAL;
+        goto error;
```
Comment (garlick) [inline src/modules/job-manager/event.c]: I think you may get into trouble with unbalanced `va_start()` / `va_end()` here?

File: src/modules/job-manager/event.c [hunk 4]
```diff
@@ -316,22 +340,37 @@ int event_job_post (struct event_ctx *ctx, struct job *job,
...
-    n = vsnprintf (context, sizeof (context), fmt, ap);
-    va_end (ap);
-    if (n < 0 || n >= sizeof (context)) {
+    if (!(o = json_vpack_ex (NULL, 0, fmt, ap))) {
         errno = EINVAL;
-        return -1;
+        goto error;
     }
-    return event_job_post (ctx, job, cb, arg, name, context);
+    if (!(context = json_dumps (o, JSON_COMPACT))) {
```
Comment (garlick) [inline src/modules/job-manager/event.c]: Need to check that encoded string doesn't exceed RFC 18 limit for context length.

File: src/modules/job-manager/submit.c [hunk 5]
```diff
@@ -107,19 +107,41 @@ zlist_t *submit_enqueue_jobs (struct queue *queue, json_t *jobs)
...
-        if (event_job_action (event_ctx, job) < 0)
-            return -1;
-        return 0;
-};
+    char event[FLUX_KVS_MAX_EVENT_CONTEXT + 1];
+    json_t *o = NULL;
+    char *str = NULL;
+    int save_errno, rv = -1;
+
+    if (!(o = json_pack ("{ s:i s:i s:i }",
```
Comment (garlick) [inline src/modules/job-manager/submit.c]: Suggestion: move event encoding to a separate function and just call `flux_kvs_event_encode_timestamp()` rather than encoding manually.

File: src/cmd/flux-job.c [hunk 6]
```diff
@@ -632,13 +642,31 @@ int cmd_id (optparse_t *p, int argc, char **argv)
...
+
 struct eventlog_ctx {
     struct flux_kvs_eventlog *log;
     optparse_t *p;
     flux_jobid_t id;
+    int format;
 };
 
-void output_event (const char *event)
+int context_format (optparse_t *p)
```
Comment (garlick) [inline src/cmd/flux-job.c]: I don't think converting to an enum is really buying us anything over a strcmp on the original option value.

File: src/cmd/flux-job.c [hunk 7]
```diff
@@ -632,13 +642,31 @@ int cmd_id (optparse_t *p, int argc, char **argv)
...
+    if (!strcasecmp (type, "text"))
+        return CONTEXT_FORMAT_TEXT;
+    else if (!strcasecmp (type, "json"))
+        return CONTEXT_FORMAT_JSON;
+    log_msg_exit ("invalid context-format type");
+    /* NEVER REACHED */
+    return -1;
+}
+
+void output_event (const char *event, int format)
```
Comment (garlick) [inline src/cmd/flux-job.c]: Suggestion: refactor to `output_event_json()` and `output_event_text()` and leave the format option checking to the main function for the subcommand?

**Comment by garlick** [comment]
I guess one question is how far we want to go with making the event context readable in "text" mode?  Can we get rid of the `note=` prefix for what used to be free form strings at the end of the context?

**Comment by grondo** [comment]
> I guess one question is how far we want to go with making the event context readable in "text" mode? Can we get rid of the note= prefix for what used to be free form strings at the end of the context?

Yeah, I had the same thought, but I'm not sure I have any good insight. For now we should probably keep things simple, but I could see a point in the future where we might want to share the per-event "decode" functions between the job-manager and the flux-job eventlog command, so that wait-event or other utilities could be extended to search and/or wait based on not only the event but the context as @chu11 had mentioned before.

(Sorry, that isn't necessarily germane to the discussion about "text" mode, except that I was thinking that perhaps each event could have a custom method for formatting...)

FWIW, it is probably fine if JSON strings are quoted in "human readable" text mode.

File: src/modules/job-manager/event.c (hunk 2 above)
Comment (chu11) [reply src/modules/job-manager/event.c]: doh, I had changed `event_job_post()` at one point, thus making the function wrapper for `post()`.

File: src/cmd/flux-job.c (hunk 6 above)
Comment (chu11) [reply src/cmd/flux-job.c]: I think I was overthinking future support, where there would be other future output formats.  It is a bit overkill.

File: src/modules/job-manager/event.c (hunk 4 above)
Comment (chu11) [reply src/modules/job-manager/event.c]: Oh yeah, in this case the check will "fallthrough" to the eventual `flux_kvs_event_encode()` call in `event_job_post()`.  I'll add a comment thought.

**Comment by chu11** [comment]
> FWIW, it is probably fine if JSON strings are quoted in "human readable" text mode.

Ok, lets go with them quoted.

**Comment by chu11** [comment]
re-pushed with suggested changes, default output of strings now includes quotes.

**Comment by garlick** [comment]
Thanks for the fixes.  Let me poke at this a bit and then I think it can go in.

@grondo, I can do the rebase of #2077 and force push, once this is merged, if that works for you.

**Comment by garlick** [comment]
This looks great @chu11, nice work as usual.   Merging.

**Comment by grondo** [comment]
> @grondo, I can do the rebase of #2077 and force push, once this is merged, if that works for you.

Do you have pending work on that branch? I'm currently adding a test, so it would be easier for me to rebase?

**Comment by garlick** [comment]
No pending work here.  If you want to do it that's OK, but I'm happy to help if needed.

**Comment by grondo** [comment]
> but I'm happy to help if needed.

Ah, now that I've started on the rebase I *totally* understand why you offered to help!
